pyanaconda/modules/payloads/payload/dnf/tree_info.py

```diff
@@ -0,0 +1,429 @@
+            proxies=proxies,
+            verify=ssl_verify,
+            cert=ssl_cert,
+            timeout=NETWORK_CONNECTION_TIMEOUT
+        )
+
+    def _download_metadata(self, downloader, url):
+        """Download metadata from the given URL."""
+        # How many times should we try the download?
+        retry_max = self.MAX_TREEINFO_DOWNLOAD_RETRIES
+
+        # Don't retry to download files that returned HTTP 404 code.
+        not_found = set()
+
+        # Retry treeinfo downloads with a progressively longer pause,
+        # so NetworkManager have a chance setup a network and we have
+        # full connectivity before trying to download things. (#1292613)
+        xdelay = xprogressive_delay()
+
+        for retry_count in range(0, retry_max):
+            # Delay if we are retrying the download.
+            if retry_count > 0:
+                log.info("Retrying download (%d/%d)", retry_count, retry_max - 1)
+                time.sleep(next(xdelay))
+
+            # Download the metadata file.
+            for name in self.TREE_INFO_NAMES:
+                file_url = "{}/{}".format(url, name)
+
+                try:
+                    with downloader(file_url) as r:
+                        if r.status_code == 404:
+                            not_found.add(name)
+
+                        r.raise_for_status()
+                        return r.text
+
+                except RequestException as e:
+                    log.debug("Failed to download '%s': %s", name, e)
+                    continue
+
+            if not_found == set(self.TREE_INFO_NAMES):
+                raise NoTreeInfoError("No treeinfo metadata found (404).")
+
+        raise NoTreeInfoError("Couldn't download treeinfo metadata.")
+
+    def verify_image_base_repo(self):
+        """Verify the base repository of an ISO image.
+
+        :return: True or False
+        """
+        repo_md = self._get_base_repository() or self._get_root_repository()
+
+        if not repo_md:
+            log.debug("There is no usable repository available")
+            return False
+
+        if not repo_md.valid:
+            log.debug("There is no valid repository available.")
+            return False
+
+        return True
+
+    def get_base_repo_url(self):
+        """Return an URL of the base repository.
+
+        :return: an URL of the base repo
+        """
+        repo_md = self._get_base_repository()
+
+        if repo_md:
+            log.debug("The treeinfo defines a base repository at: %s", repo_md.path)
+            return repo_md.path
+
+        log.debug("No base repository found in the treeinfo. Using installation tree root.")
+        return self._root_path
+
+    def _get_base_repository(self):
+        """Return metadata of the base repository.
+
+        :return: an instance of TreeInfoRepoMetadata or None
+        """
+        repos = DEFAULT_REPOS
+
+        for repo_md in self.repositories:
+            if repo_md.name in repos:
```

Comment:
I think this constant could be in-lined. It would be even more readable that way.

Reply:
Fixed.